[bug]: TS < 4.8 project no longer compiles with spring ~9.5

### Which react-spring target are you using?

- [X] `@react-spring/web`
- [ ] `@react-spring/three`
- [ ] `@react-spring/native`
- [ ] `@react-spring/konva`
- [ ] `@react-spring/zdog`

### What version of react-spring are you using?

9.5.4

### What's Wrong?

Due to the changes in #196, backwards compatibility appears to break. Trying to compile the project on TS 4.2.4 gives ~40 errors from react-spring
```
> node node_modules/.bin/tsc --project tsconfig.json --noEmit --pretty

node_modules/@react-spring/core/dist/declarations/src/hooks/useSpring.d.ts:8:116 - error TS1005: '?' expected.

8 export declare type UseSpringProps<Props extends object = any> = unknown & PickAnimated<Props> extends infer State extends Lookup<any> ? Remap<ControllerUpdate<State> & {
                                                                                                                     ~~~~~~~

node_modules/@react-spring/core/dist/declarations/src/hooks/useSpring.d.ts:8:124 - error TS1005: ':' expected.

8 export declare type UseSpringProps<Props extends object = any> = unknown & PickAnimated<Props> extends infer State extends Lookup<any> ? Remap<ControllerUpdate<State> & {                                                                                                                        ~~~~~~
```

I think this was unintentional as the fix was to support TS 4.8 (not remove support for older TS), but perhaps should have been released as a major version. We're on 4.2.4 and can no longer compile our project with these changes. Additionally, because react-spring has dependencies set to `~9.5.2`, all new installs will also pull 9.5.4 meaning people must peg all dependencies in their package files:
```
    "@react-spring/types": "9.5.2",
    "@react-spring/web": "9.5.2",
    "@react-spring/animated": "9.5.2",
    "@react-spring/core": "9.5.2",
    "@react-spring/shared": "9.5.2",
```
NOTE: I did not test other TS versions between 4.8, so it's possible that it works on 4.3 even

### To Reproduce

- Add react-spring 9.5.2 to a project using ts 4.2.3

### Expected Behaviour

react-spring should be compilable or have specified TS dependency ranges.

Various workarounds are likely possible, but the ergonomics are a little funky.

### Link to repo

Don't have

--- a/.changeset/config.json
+++ b/.changeset/config.json
@@ -6,5 +6,5 @@
   "access": "public",
   "baseBranch": "master",
   "updateInternalDependencies": "patch",
-  "ignore": []
+  "ignore": ["@react-spring/demo", "@react-spring/docs"]
 }

--- a/.changeset/cuddly-dolls-sit.md
+++ b/.changeset/cuddly-dolls-sit.md
@@ -1,0 +1,16 @@
+---
+'@react-spring/core': patch
+'@react-spring/animated': patch
+'@react-spring/parallax': patch
+'@react-spring/rafz': patch
+'react-spring': patch
+'@react-spring/shared': patch
+'@react-spring/types': patch
+'@react-spring/konva': patch
+'@react-spring/native': patch
+'@react-spring/three': patch
+'@react-spring/web': patch
+'@react-spring/zdog': patch
+---
+
+fix: allow transpile from TS < 4.7

--- a/.changeset/pre.json
+++ b/.changeset/pre.json
@@ -1,0 +1,21 @@
+{
+  "mode": "pre",
+  "tag": "beta",
+  "initialVersions": {
+    "@react-spring/demo": "0.0.1",
+    "@react-spring/docs": "1.0.7",
+    "@react-spring/animated": "9.5.4",
+    "@react-spring/core": "9.5.4",
+    "@react-spring/parallax": "9.5.4",
+    "@react-spring/rafz": "9.5.4",
+    "react-spring": "9.5.4",
+    "@react-spring/shared": "9.5.4",
+    "@react-spring/types": "9.5.4",
+    "@react-spring/konva": "9.5.4",
+    "@react-spring/native": "9.5.4",
+    "@react-spring/three": "9.5.4",
+    "@react-spring/web": "9.5.4",
+    "@react-spring/zdog": "9.5.4"
+  },
+  "changesets": []
+}

--- a/packages/core/src/hooks/useSpring.ts
+++ b/packages/core/src/hooks/useSpring.ts
@@ -10,17 +10,19 @@ import { useSprings } from './useSprings'
  * The props that `useSpring` recognizes.
  */
 export type UseSpringProps<Props extends object = any> = unknown &
-  PickAnimated<Props> extends infer State extends Lookup<any>
-  ? Remap<
-      ControllerUpdate<State> & {
-        /**
-         * Used to access the imperative API.
-         *
-         * When defined, the render animation won't auto-start.
-         */
-        ref?: SpringRef<State>
-      }
-    >
+  PickAnimated<Props> extends infer State
+  ? State extends Lookup
+    ? Remap<
+        ControllerUpdate<State> & {
+          /**
+           * Used to access the imperative API.
+           *
+           * When defined, the render animation won't auto-start.
+           */
+          ref?: SpringRef<State>
+        }
+      >
+    : never
   : never
 
 /**
@@ -32,8 +34,10 @@ export function useSpring<Props extends object>(
     | Function
     | (() => (Props & Valid<Props, UseSpringProps<Props>>) | UseSpringProps),
   deps?: readonly any[] | undefined
-): PickAnimated<Props> extends infer State extends Lookup<any> 
-  ? [SpringValues<State>, SpringRef<State>]
+): PickAnimated<Props> extends infer State
+  ? State extends Lookup
+    ? [SpringValues<State>, SpringRef<State>]
+    : never
   : never
 
 /**
@@ -49,8 +53,10 @@ export function useSpring<Props extends object>(
 export function useSpring<Props extends object>(
   props: (Props & Valid<Props, UseSpringProps<Props>>) | UseSpringProps,
   deps: readonly any[] | undefined
-): PickAnimated<Props> extends infer State extends Lookup<any>
-  ? [SpringValues<State>, SpringRef<State>]
+): PickAnimated<Props> extends infer State
+  ? State extends Lookup
+    ? [SpringValues<State>, SpringRef<State>]
+    : never
   : never
 
 /** @internal */

--- a/packages/core/src/hooks/useSprings.ts
+++ b/packages/core/src/hooks/useSprings.ts
@@ -43,8 +43,10 @@ export function useSprings<Props extends UseSpringProps>(
   length: number,
   props: (i: number, ctrl: Controller) => Props,
   deps?: readonly any[]
-): PickAnimated<Props> extends infer State extends Lookup<any>
-  ? [SpringValues<State>[], SpringRefType<State>]
+): PickAnimated<Props> extends infer State
+  ? State extends Lookup<any>
+    ? [SpringValues<State>[], SpringRefType<State>]
+    : never
   : never
 
 /**
@@ -62,8 +64,10 @@ export function useSprings<Props extends UseSpringsProps>(
   length: number,
   props: Props[] & UseSpringsProps<PickAnimated<Props>>[],
   deps: readonly any[] | undefined
-): PickAnimated<Props> extends infer State extends Lookup<any>
-  ? [SpringValues<State>[], SpringRefType<State>]
+): PickAnimated<Props> extends infer State
+  ? State extends Lookup<any>
+    ? [SpringValues<State>[], SpringRefType<State>]
+    : never
   : never
 
 /** @internal */

--- a/packages/core/src/hooks/useTrail.ts
+++ b/packages/core/src/hooks/useTrail.ts
@@ -19,8 +19,10 @@ export function useTrail<Props extends object>(
     ctrl: Controller
   ) => UseTrailProps | (Props & Valid<Props, UseTrailProps<Props>>),
   deps?: readonly any[]
-): PickAnimated<Props> extends infer State extends Lookup<any>
-  ? [SpringValues<State>[], SpringRef<State>]
+): PickAnimated<Props> extends infer State
+  ? State extends Lookup<any>
+    ? [SpringValues<State>[], SpringRef<State>]
+    : never
   : never
 
 export function useTrail<Props extends object>(
@@ -32,8 +34,10 @@ export function useTrail<Props extends object>(
   length: number,
   props: UseTrailProps | (Props & Valid<Props, UseTrailProps<Props>>),
   deps: readonly any[]
-): PickAnimated<Props> extends infer State extends Lookup<any>
-  ? [SpringValues<State>[], SpringRef<State>]
+): PickAnimated<Props> extends infer State
+  ? State extends Lookup<any>
+    ? [SpringValues<State>[], SpringRef<State>]
+    : never
   : never
 
 export function useTrail(

--- a/packages/core/src/hooks/useTransition.tsx
+++ b/packages/core/src/hooks/useTransition.tsx
@@ -45,8 +45,10 @@ export function useTransition<Item, Props extends object>(
     | UseTransitionProps<Item>
     | (Props & Valid<Props, UseTransitionProps<Item>>),
   deps?: any[]
-): PickAnimated<Props> extends infer State extends Lookup
-  ? [TransitionFn<Item, PickAnimated<Props>>, SpringRefType<State>]
+): PickAnimated<Props> extends infer State
+  ? State extends Lookup
+    ? [TransitionFn<Item, PickAnimated<Props>>, SpringRefType<State>]
+    : never
   : never
 
 export function useTransition<Item, Props extends object>(
@@ -62,8 +64,10 @@ export function useTransition<Item, Props extends object>(
     | UseTransitionProps<Item>
     | (Props & Valid<Props, UseTransitionProps<Item>>),
   deps: any[] | undefined
-): PickAnimated<Props> extends infer State extends Lookup
-  ? [TransitionFn<Item, State>, SpringRefType<State>]
+): PickAnimated<Props> extends infer State
+  ? State extends Lookup
+    ? [TransitionFn<Item, State>, SpringRefType<State>]
+    : never
   : never
 
 export function useTransition(

